text/sigma-protocol/README.md

```diff
@@ -0,0 +1,76 @@
+Being able to prove such statements is important for anonymous credentials because in those the credential signature is of a form Zg<sub>1</sub><sup>`m1`</sup>g<sub>2</sub><sup>`m2`</sup>g<sub>3</sub><sup>`m3`</sup>...g<sub>n</sub><sup>`n`</sup>h<sup>`r`</sup> where `Z` is derived from signer's secret key and each of m<sub>i</sub> is a credential attribute and `r` is the randomness. There might be more elements in the signature but the aforementioned is of interest here. When using such credentials for zero knowledge proofs, the prover proves knowledge of the signature and the attributes in there. Depending on the signature scheme used, the prover randomizes the signature in a particualar way before creating a proof of knowledge of the messages and signature. This RFC will give examples in the context of PS signatures which are implemented [here](https://github.com/lovesh/signature-schemes/tree/master/ps).  
+Interactive Sigma protocols are 3-step, 
+- in the first step, prover commits to some randomness and sends it to the verifier.
+- in 2nd step, prover gets a challenge from the verifier.
+- in 3rd step, prover creates a response involving the randomness, its secret and the challenge and sends to the verifier.
+- the verifier will now check correctness of a relation involving the commitment from 1st step, challenge and response.
+
+Non interactive Sigma protocols modify the 2nd step. Rather than getting the challenge from the verifier, they compute the challenge by hashing the commitments from 1st step.   
+Consider an example of prover proving knowledge of `x` in g<sup>`x`</sup>. Lets call g<sup>`x`</sup> as Y. Only the prover knows `x` but both prover and verifier know `Y` and `g`. 
+- prover picks a random number `r`, creates commitment T = g<sup>`r`</sup> and sends T to the verifier.
+- verifier sends challenge c to prover.
+- prover computes response `s = r - c*x` and sends `s` to verifier.
+- verifier checks if g<sup>`s`</sup>*Y<sup>`c`</sup> equals T. If it does then verifier is convinced of prover's knowledge of `x`.
+
+In the non-interactive version prover will hash g, Y and T to compute challenge c. There are more details and variations for the above protocol which can be seen [here](https://en.wikipedia.org/wiki/Proof_of_knowledge#Sigma_protocols).  Similar protocols can be developed more more complex relations that can be seen [here](https://medium.com/@loveshharchandani/zero-knowledge-proofs-with-sigma-protocols-91e94858a1fb).
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This section will describe how such proofs of knowledge are to be constructed in reference to [PS signatures](https://eprint.iacr.org/2015/525), 6.2. The code for the proofs of knowledge are [here](https://github.com/lovesh/signature-schemes/blob/master/ps/src/pok.rs).  
```

Comment:
I think the code should be included with the RFC, not stay in a personal repo.

Reply:
We decided in the Ursa meeting that it would be fine to submit this code for merging in Ursa so i will move the code. But point noted for future submissions.